modules/tensor_mechanics/src/materials/ComputeLinearElasticPFFractureStress.C

```diff
@@ -8,112 +8,247 @@
+{
+  // Isotropic elasticity is assumed and should be enforced
+  const Real lambda = _elasticity_tensor[_qp](0, 0, 1, 1);
+  const Real mu = _elasticity_tensor[_qp](0, 1, 0, 1);
+  const Real k = lambda + 2.0 * mu / LIBMESH_DIM;
+
+  RankTwoTensor I2(RankTwoTensor::initIdentity);
+  RankFourTensor I2I2 = I2.outerProduct(I2);
+
+  RankFourTensor Jacobian_pos, Jacobian_neg;
+  RankTwoTensor strain0vol, strain0dev;
+  RankTwoTensor stress0pos, stress0neg;
+  Real strain0tr, strain0tr_neg, strain0tr_pos;
+
+  strain0dev = _mechanical_strain[_qp].deviatoric();
+  strain0vol = _mechanical_strain[_qp] - strain0dev;
+  strain0tr = _mechanical_strain[_qp].trace();
+  strain0tr_neg = std::min(strain0tr, 0.0);
+  strain0tr_pos = strain0tr - strain0tr_neg;
+  stress0neg = k * strain0tr_neg * I2;
+  stress0pos = _elasticity_tensor[_qp] * _mechanical_strain[_qp] - stress0neg;
+  // Energy with positive principal strains
+  RankTwoTensor strain0dev2 = strain0dev * strain0dev;
+  F_pos = 0.5 * k * strain0tr_pos * strain0tr_pos + mu * strain0dev2.trace();
+  F_neg = 0.5 * k * strain0tr_neg * strain0tr_neg;
+
+  _stress[_qp] = stress0pos * _D[_qp] + stress0neg;
 
   // 2nd derivative wrt c and strain = 0.0 if we used the previous step's history varible
   if (_use_current_hist)
-    _d2Fdcdstrain[_qp] = _dstress_dc[_qp];
+    _d2Fdcdstrain[_qp] = stress0pos * _dDdc[_qp];
+
+  // Used in StressDivergencePFFracTensors off-diagonal Jacobian
+  _dstress_dc[_qp] = stress0pos * _dDdc[_qp];
+
+  if (strain0tr < 0)
+    Jacobian_neg = k * I2I2;
+  Jacobian_pos = _elasticity_tensor[_qp] - Jacobian_neg;
+  _Jacobian_mult[_qp] = _D[_qp] * Jacobian_pos + Jacobian_neg;
+}
+
+void
+ComputeLinearElasticPFFractureStress::computeQpStress()
+{
+  Real F_pos, F_neg;
+
+  switch (_decomposition_type)
+  {
+    case 0:
+      computeStrainSpectral(F_pos, F_neg);
+      break;
+    case 1:
+      computeStrainVolDev(F_pos, F_neg);
+      break;
+    case 2:
+      computeStressSpectral(F_pos, F_neg);
+      break;
+    default:
+    {
+      _stress[_qp] = _D[_qp] * _elasticity_tensor[_qp] * _mechanical_strain[_qp];
+      F_pos = (_stress[_qp]).doubleContraction(_mechanical_strain[_qp]) / 2.0;
+      F_neg = 0.0;
+      if (_use_current_hist)
+        _d2Fdcdstrain[_qp] = _stress[_qp] * _dDdc[_qp];
+
+      _dstress_dc[_qp] = _stress[_qp] * _dDdc[_qp];
+      _Jacobian_mult[_qp] = _D[_qp] * _elasticity_tensor[_qp];
+    }
+  }
+
+  // // Assign history variable
+  if (F_pos > _H_old[_qp])
+    _H[_qp] = F_pos;
+  else
+    _H[_qp] = _H_old[_qp];
+
+  Real hist_variable = _H_old[_qp];
+  if (_use_current_hist)
+    hist_variable = _H[_qp];
+
+  if (hist_variable < _barrier[_qp])
+    hist_variable = _barrier[_qp];
 
-  // Assign L and kappa
-  _kappa[_qp] = _gc_prop[_qp] * _l[_qp];
-  _L[_qp] = 1.0 / (_gc_prop[_qp] * _visco[_qp]);
+  // Elastic free energy density
+  _E[_qp] = hist_variable * _D[_qp] + F_neg;
```

Comment:
I think _E[_qp] does not have the contribution from F_neg.
hist_variable * _D[_qp] + F_neg is indeed the elastic energy, but I think what enters the phase field addition energy is only the positive part of this elastic energy. Another way to think about this is that fracture starts to grow once _E exceeds _barrier, therefore _E should be only the positive strain energy.

Reply:
If we don't need F_neg, then we don't need to compute F_neg at all in all the decompositions, which saves computational cost by almost a half.

Reply:
Since we calculate elastic free energy density, not just driving energy, I would prefer to keep F_neg here.